src/main/java/spoon/reflect/visitor/chain/QueryStep.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright (C) 2006-2016 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.visitor.chain;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import spoon.Launcher;
+import spoon.SpoonException;
+import spoon.reflect.visitor.filter.Scann;
+
+/**
+ * A) scan(Filter) scan all child elements of input element and send to output only these elements, which matches the filter
+ * B) then(AsyncFnc) initialize filter with input element and then scan all children of the start element returned by filter and send to output only these elements, which matches the filter
+ * C) scan all children of the start element returned by filter and send to output only these elements, which matches the filter - ignore input
+ * D) matches(Predicate) - send input to output if it matches filter
+ *
+ *
+ * @param <I> - input type
+ * @param <O> - output type
+ */
+public abstract class QueryStep<O> implements Consumer<Object> {

```

Comment:
what about having an interface `QueryStep`? (in Spoon, the design is very interface-oriented: most key concepts and key methods are defined in interfaces). The method `then` seems indeed essential for the design.